src/reflect/scala/reflect/internal/Symbols.scala

```diff
@@ -3769,20 +3763,36 @@ trait Symbols extends api.Symbols { self: SymbolTable =>
   }
 
   /** A class for type histories */
-  private final case class TypeHistory(var validFrom: Period, info: Type, prev: TypeHistory) {
+  private final case class TypeHistory private (private var _validFrom: Period, private var _info: Type, private var _prev: TypeHistory) {
+    def validFrom: Period = _validFrom
+    def info: Type = _info
+    def prev: TypeHistory = _prev
+
     assert((prev eq null) || phaseId(validFrom) > phaseId(prev.validFrom), this)
     assert(validFrom != NoPeriod, this)
 
+    // OPT: mutate the current TypeHistory rather than creating a new one. TypeHistory instances should not be shared.
+    final def reset(validFrom: Period, info: Type): TypeHistory = {
+      if (this ne noTypeHistory) TypeHistory(validFrom, info, null)
```

Comment:
I think this has the branches switched... because `NoTypeHistory` is a constant to be shared, that is the one that should never be mutated.